pytorch/pytorch_ddp_1machine.py

```diff
@@ -0,0 +1,216 @@
+"""
+Optuna example that optimizes multi-layer perceptrons using PyTorch distributed.
+
+In this example, we optimize the validation accuracy of fashion product recognition using
+PyTorch distributed data parallel and FashionMNIST.
+This example is dedicated to a use case of using one machine with multiple GPUs
+where no MPI is needed.
+We optimize the neural network architecture as well as the optimizer configuration.
+As it is too time consuming to use the whole FashionMNIST dataset, we
+here use a small subset of it.
+
+You can execute this example with a command as follows:
+    $ python pytorch_ddp_1machine.py
+
+"""
+
+from functools import partial
+import os
+
+import optuna
+from optuna.trial import TrialState
+import torch
+import torch.distributed as dist
+import torch.multiprocessing as mp
+import torch.nn as nn
+import torch.nn.functional as F
+from torch.nn.parallel import DistributedDataParallel as DDP
+import torch.optim as optim
+import torch.utils.data
+from torchvision import datasets
+from torchvision import transforms
+
+
+# Arbitrary GPUs can be selected
+DEVICE_IDS = [0, 1]
```

Comment:
Due to this part, this example is not `arbitrary multiple GPUs.` I believe that `torchrun` style script would be more practical in my PyTorch experience.

Reply:
By `arbiturary multiple GPUs`, I mean that you can specify exactly which GPUs to use. I understand that in a large scale setting where you have multiple nodes that have multiple GPUs in each node, using MPI (and I guess as well as `torchrun` because it seems to require MPI) would be definitely a better practice. 

This example, however, was made with the mind that you don't need to install any MPI packages. Say, I have a local machine with 4 GPUs and I want to run an experiment on GPU1 & 3 because I want GPU0 to be dedicated to graphics and GPU2 is running other programs. I can just run the code as a normal python code like `python pytorch/pytorch_ddp_1machine.py`. 

I can make `DEVICE_IDS` an argument so that user can specify which particular GPUs to use from command line.

Reply:
Thanks. I got the point. I should have discussed the necessity of this PR because https://github.com/optuna/optuna-examples/pull/150 also addresses the same issue.

In terms of the number of gpus, we can manage it from the command line; For example, https://discuss.pytorch.org/t/how-to-change-the-default-device-of-gpu-device-ids-0/1041/2.

Reply:
>In terms of the number of gpus, we can manage it from the command line; For example, https://discuss.pytorch.org/t/how-to-change-the-default-device-of-gpu-device-ids-0/1041/2.

Do you mean passing the environment variables in the command line like this?
`CUDA_VISIBLE_DEVICES=1,2 python myscript.py`

I was told changing environment variable is generally not a good practice, so I tried avoiding this. Also, with the above example, I guess you might need to configure a default value in the python code.
In the updated commit, I offered a suggestion using `--device_ids` argument to specify `device_ids` with a default value 0. Let me know if you like it or what you have in your mind.

Reply:
> Do you mean ...
Yes!

> I was told changing environment variable is generally not a good practice

Could you share any reference, ideally, PyTorch page? Thanks!

Reply:
>Could you share any reference, ideally, PyTorch page? Thanks!

I can't find a good reference but I feel the idea comes from more production or security side after some google search.
To summarize, my understanding is that

- Potentially making debugging harder because the environment change may not be applied immediately.
https://support.cloud.engineyard.com/hc/en-us/articles/205407508-Environment-Variables-and-Why-You-Shouldn-t-Use-Them
- There're risks of leaking sensitive information
https://blog.dnsimple.com/2018/05/using-environment-as-configuration/

Again, I was just told by a colleague of mine and I don't really have any experience about it. And apparently, there also seem to be many cases that env variables should be actively used. So I don't have strong opinion about this.
I just thought if there was a less risky way, then I just pick that. 

So, if you think using environment variable is better, then I can change the code accordingly.